### PR TITLE
uv: support for android build

### PR DESCRIPTION
**Pull request by linusmartensson**
Adds targets and fixes to ensure building for android works.

To build on android, refer to the android standalone toolchain.

### CODE REVIEW

**Comment by lucab** [comment]
Hi, as part of rust we did the same android porting; the code is here https://github.com/lucab/libuv/commit/f71151f54aabef0be62b858ca239775326b4a88d
I was going to submit a PR, when spotted this one and didn't proceed with mine.

Looking at the code, it looks like it is almost identical (yours also a bit cleaner), so I'll probably drop mine. 
What's the status of having this PR merged? (So that rust can go back to mainline libuv...)

File: src/unix/linux-syscalls.h [hunk 1]
```diff
@@ -28,6 +28,9 @@
 #include <stdint.h>
 #include <signal.h>
 #include <sys/types.h>
+#ifdef ANDROID
+#include <time.h>
```
Comment (bnoordhuis) [inline src/unix/linux-syscalls.h]: I'm guessing that's for struct timespec? You can include it unconditionally, I was relying on timespec being pulled in implicitly.

File: src/unix/linux-core.c [hunk 2]
```diff
@@ -36,7 +36,10 @@
 #include <fcntl.h>
 #include <time.h>
 
+#ifndef ANDROID
 #define HAVE_IFADDRS_H 1
+#endif
```
Comment (bnoordhuis) [inline src/unix/linux-core.c:41]: Doesn't android have ifaddrs.h?

EDIT: Okay, looks like it doesn't. I guess we could implement getifaddrs() ourselves through ioctl(SIOCGIFCONF) and AF_NETLINK sockets but meh, I don't think many people care. I'll wait for the bug reports to come in first. :-)

File: src/unix/stream.c [hunk 3]
```diff
@@ -705,8 +705,20 @@ static void uv__write(uv_stream_t* stream) {
   iovcnt = req->bufcnt - req->write_index;
 
   /* Limit iov count to avoid EINVALs from writev() */
+#if defined(ANDROID)
+  /*
+   * Android does not have this value predefined,
+   * we need to retrieve it using the syscnf interface.
```
Comment (bnoordhuis) [inline src/unix/stream.c]: That sounds unlikely. Some quick grepping through an android source tree suggests it's there. Could it be there's a compiler flag or define hiding it?

File: src/unix/thread.c [hunk 4]
```diff
@@ -26,9 +26,9 @@
 #include <assert.h>
 #include <errno.h>
 
-#if defined(__APPLE__) && defined(__MACH__)
+#if (defined(__APPLE__) && defined(__MACH__)) || defined(ANDROID)
 #include <sys/time.h>
-#endif /* defined(__APPLE__) && defined(__MACH__) */
+#endif /* (defined(__APPLE__) && defined(__MACH__)) || defined(ANDROID) */
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Feel free to include that header unconditionally. I'm not sure why it's wrapped in an #ifdef guard in the first place.

**Comment by bnoordhuis** [comment]
Couple of nits but mostly LGTM. If you fix them up, I'll land it.

File: src/unix/stream.c (hunk 3 above)
Comment (linusmartensson) [inline src/unix/stream.c]: Unfortunately, it's not available in the bionic version of libc at all, which is what we use from the NDK. The alternative to fetching it dynamically would be to provide a default value on ifndef IOV_MAX, but I felt a dynamic fetch via sysconf was more flexible.

There's references to a default definition in some prebuilt components here, but this is as far as I can tell a fallback system for gcc internals:
http://androidxref.com/4.2.2_r1/search?q=IOV_MAX&defs=&refs=&path=&hist=&project=prebuilts

Other external packages that rely on the variable define it themselves:
http://androidxref.com/4.2.2_r1/search?q=IOV_MAX&defs=&refs=&path=&hist=&project=external

Comment (bnoordhuis) [inline src/unix/stream.c]: Hm, okay. Can you move this code into a separate function called e.g. getiovmax? I suggest replacing the `#if defined(ANDROID)` with `#if !defined(IOV_MAX)`.

File: src/unix/linux-core.c (hunk 2 above)
Comment (linusmartensson) [inline src/unix/linux-core.c:41]: Double-checked, and no, it's not available in the SDK. The dhcpd-implementation in the android sources (http://androidxref.com/4.2.2_r1/xref/external/dhcpcd/) provides it's own implementation of ifaddrs based on traversal of sysfs. I presume it could be copied to NodeJS under the BSD/Apache license referenced in that library, but this should probably be a separate patch.

File: src/unix/stream.c (hunk 3 above)
Comment (piscisaureus) [inline src/unix/stream.c]: Would it be better to make `iovmax` of type `sig_atomic_t` ?

Comment (bnoordhuis) [inline src/unix/stream.c]: Why? Because of concurrent loads / stores?  That doesn't really matter, at worst you make a few superfluous calls to sysconf().  Besides, sig_atomic_t is an int on all platforms and architectures. :-)

File: src/unix/thread.c (hunk 4 above)
Comment (linusmartensson) [inline src/unix/thread.c]: Done in latest patch version.

File: src/unix/stream.c (hunk 3 above)
Comment (linusmartensson) [inline src/unix/stream.c]: Moved to separate function, and I added a final fallback on 1024 if neither _SC_IOV_MAX_ nor IOV_MAX was available.

File: src/unix/pthread-fixes.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+   This contains Pthread-related functions not provided by the Android NDK
+   but required by the Breakpad unit test. The functions are inlined here
+   in a C++ anonymous namespace in order to keep the build files simples.
```
Comment (piscisaureus) [inline src/unix/pthread-fixes.c]: Remove the comment about C++ namespace and breakpad?

File: uv.gyp [hunk 6]
```diff
@@ -20,7 +20,8 @@
         'conditions': [
           ['OS=="solaris"', {
             'cflags': [ '-pthreads' ],
-          }, {
+          }],
+          ['OS!="android"', {
```
Comment (bnoordhuis) [inline uv.gyp:24]: Just realized that this breaks the sunos gyp build. I'll see if I can fix it up.

**Comment by bnoordhuis** [comment]
Thanks Linus, landed with some fix-ups in fc6a2ad.
